### ISSUE TITLE
support dynamic values

### ISSUE THREAD

**Issue by VinSpee** [picostyle/picostyle#14]
Hi!

I authored and maintain a library called [`react-shed`](https://github.com/tedconf/react-shed/tree/emotion), which uses a css-in-js library to allow users to define styles that are then applied to elements like this:

```jsx
<Shed
  component="div"
  mx="5"
>output is margin-left: sizes(5), margin-right: sizes(5)</Shed>
```

is is possible to calculate values based on props with _picostyle_?

Thanks!

**Comment by jbucaran** [comment]
Hi @VinSpee, 

How do you think that would work with picostyle? Can you write some pseudo code? The snippet posted above does not seem familiar to me.

**Comment by VinSpee** [comment]
Sure - something like this:

```js
const MyH1 = pico(‘h1’)(propsPassedToPico => ({
  marginLeft: getSize(propsPassedToPico.mx),
  marginRight: getSize(propsPassedToPico.mx),
});
```

```jsx
<MyH1 mx=“1”>has a margin left and right of whatever getSize(‘1’) is</MyH1>
```

**Comment by jbucaran** [comment]
@VinSpee Did you mean?

```js
<MyH1 mx="1">has a margin left and right of whatever {getSize("1")} is</MyH1>
```

And one more thing, just to clarify: we are _not_ talking about adding `px` to numbers by default if no unit is given right?

**Comment by VinSpee** [comment]
- no, I am _not_ referring to adding `px` by default. 

The definition of `getSize` and `mx` are arbitrary. Another example could look like this:

```jsx
const MyButton = pico(‘button’)(propsPassedToPico => ({
  fontSize: propsPassedToPico.big ? ‘4rem’ : ‘1rem’,
});
```

And usage like this:

```jsx
<MyButton big>big button</MyButton>
```

Does that make sense?

**Comment by jbucaran** [comment]
@VinSpee Hmm, okay, I think I got it now. Picostyle doesn't do AFIK, but hey, that looks pretty interesting.

**Comment by morishitter** [comment]
@VinSpee Thank you for your proposal 😉 styled-components has this feature too. IMO, we don't use CSS-in-JS libraries for reusability but for deletability so I'd not like to use it. If we implement its function, we will be able to god picostyle component like the following code.

```js
const GodDiv = picostyle('div')(props => ({
  fontSize: props.fs + 'px',
  color: props.color,
  liheHeight: props.lh,
  // ...
})
```

```js
<GodDiv fs=14 color="tomato" lh=1>It seems like using utility classes. Why we use CSS-inJS? 🤔</GodDiv>
```

This may be an extreme example, but functions induce our actions. Picostyle is simple so it's lightweight. I'd like to keep it.

**Comment by JorgeBucaran** [comment]
@morishitter 👍 

Do you think we could still implement this via a HOF of `picostyle`? Just like `picostyle` is a HOC of `hyperapp.h` or `picodom.h`.

**Comment by morishitter** [comment]
@JorgeBucaran Yes, If we implement to support dynamic values, it is as a HOC :)

**Comment by VinSpee** [comment]
@morishitter:
> This may be an extreme example

Yes, but what's the problem with it?

> It seems like using utility classes. Why we use CSS-inJS?

Yes, for many reasons, I outline a lot of them here: https://speakerdeck.com/vinspee/im-over-css-function-atomic-css-for-the-masses

TL;DR: quick assignment of values from a predefined theme, ex

`c="blue"` outputs `color: #2EA2FC`
`f="3"` outputs `fontSize: 4.2333rem`

etc.

**Comment by JorgeBucaran** [comment]
@VinSpee Do you mean **built-into** the library? What if we push it into userland instead. It should be easy to implement no?

**Comment by VinSpee** [comment]
no no no no notu built-in to the library at all, but available in userland so I _can_ implement something like this yes.

**Comment by JorgeBucaran** [comment]
> @VinSpee ...available in userland so I can implement something like this yes.

Yes, that's what I mean too. So, what would need to change in picostyle for anyone to be able to implement something like this? I thought it was possible already.

**Comment by VinSpee** [comment]
this _can_ be done in userland like this:

```js
  const Text = (props, children) => ps("h1")({
    fontSize: "64px",
    lineHeight: "1.5",
    cursor: "pointer",
    color: props.c,
  })(props, children);

<Text c="cyan">This will be cyan</Text>
```

**Comment by IamCarbonMan** [comment]
Perhaps I'm just reading the code wrong, but you use `sheet.insertRule(rule, 0)` to write the compiled rules to the document. This means that if I have, for example, a button that changes color and size when you click it, and the user clicks that button 20 times, then 20 new rules will be added to the sheet. Since they all have the same selector, the last one found in the sheet will have the highest specificity. How does this work then?
